packages/parser/src/parser.ts

```diff
@@ -0,0 +1,59 @@
+import { HandlerMethodDecorator } from '@aws-lambda-powertools/commons/types';
+import { Context, Handler } from 'aws-lambda';
+import { ZodSchema } from 'zod';
+import { type ParserOptions } from './types/ParserOptions.js';
+
+/**
+ * A decorator to parse your event.
+ *
+ * @example
+ * ```typescript
+ *
+ * import { parser } from '@aws-lambda-powertools/parser';
+ * import { sqsEnvelope } from '@aws-lambda-powertools/parser/envelopes/sqs';
+ *
+ *
+ * const Order = z.object({
+ *  orderId: z.string(),
+ *  description: z.string(),
+ * }
+ *
+ * class Lambda extends LambdaInterface {
+ *
+ *   @parser({ envelope: sqsEnvelope, schema: OrderSchema })
+ *   public async handler(event: Order, _context: Context): Promise<unknown> {
+ *   // sqs event is parsed and the payload is extracted and parsed
+ *   // apply business logic to your Order event
+ *   const res = processOrder(event);
+ *   return res;
+ *   }
+ * }
+ *
+ * @param options
+ */
+const parser = <S extends ZodSchema>(
+  options: ParserOptions<S>
+): HandlerMethodDecorator => {
+  return (_target, _propertyKey, descriptor) => {
+    const original = descriptor.value!;
+
+    const { schema, envelope } = options;
+
+    descriptor.value = function (
```

Comment:
Does this still keep the decorated method async? I haven't stepped through with the debugger, but I think with this implementation we are implicitly replacing an async method with a sync one and when calling it with the `apply()` we are returning an unresolved promise.

Not sure if this is an issue but I think this is why we made all other decorators async.

What do you think?

Reply:
Great observation! I checked in the debugger and the call `original.apply()` remains async, while `descriptor.value` is of course sync function and thus all the method of the class will be replaced with a sync function, changing return type value instead of Promise.

We could check `original.constructor.name === AsyncFunction` to decide what we should return. The more interesting case will bubble up once users want to bring their custom `transforms` or `refines`  which include async code. This would force us to expose `parseAsync` and branch the parsing for envelopes and decorator/middy. 

I'd suggest to change the modified `descriptor.value` to async, to keep it consistent with other decorator and we will circle back in a separate issue on this matter. We planned to address the sync/async decorators anyway.

Reply:
Yea I agree that we should stick to async and keep things simple and consistent.

We discussed doing runtime checks for the constructor type to determine if it's sync or async when we implemented the Batch Processor and we eventually decided to drop the idea.

The main argument against this is that even though within the function body you might be able to do this check, the function signature will always have to be `async`, which then leads to the return type always being a `Promise<T>`.